### PR TITLE
Implemented block params support for `each` and `with`

### PR DESCRIPTION
**Pull request by ericbn**
Block parameter does not change current context

### CODE REVIEW

**Comment by kpdecker** [comment]
These look like large changes to existing features. Can you provide a summary of the changed and new behaviors?

**Comment by ericbn** [comment]
Sure, @kpdecker.

Summary of changes:
* Implemented block params support for `with` block (this was requested in #1042)
* Changed current block params support implementation for `each` to make it "context-preserving". The same context-preserving behavior applies to the `with` with block params implemented above.
* Fixed the `each` helper so it also iterates over an empty string property in objects (as requested in #1021).
* Added unit tests for all corrections and new implementations.

Now, the context-preserving behavior for blocks with block parameters comes from the Ember specifications of the block params feature. Refences about this can be found in the corresponding [RFC](https://github.com/emberjs/rfcs/pull/3#issuecomment-61543805):

>  Introduce block parameters to the Handlebars language to standardize context-preserving helpers ([ref](https://github.com/emberjs/rfcs/pull/3#issue-40512558))

This idea of a context-preserving block was already introduced to Ember in their custom `{{#each ... in ...}}` and `{{#with ... as ...}}`: ", and the generic block param specification aims to make it standard:

 > `{{#each person in persons}}` was designed to not lose the parent context. It works, but its a hack which should be standardized ([ref](https://github.com/emberjs/rfcs/pull/3#issuecomment-57810150))

The [Ember 1.10 release notes](http://emberjs.com/blog/2015/02/07/ember-1-10-0-released.html#toc_block-params) also gives a summary about it.

File: spec/builtins.js [hunk 1]
```diff
@@ -47,6 +47,22 @@ describe('builtin helpers', function() {
       var string = '{{#with person}}Person is present{{else}}Person is not present{{/with}}';
       shouldCompileTo(string, {}, 'Person is not present');
     });
```
Comment (ericbn) [inline spec/builtins.js]: @kpdecker, I hope the specs can also serve as a more detailed summary of the changes.

**Comment by ericbn** [comment]
There is a change I did in the `each` helper, when block params are used, that deserves also a explanation.

The block params for `each` are not only the iteration item (first param), but also key, index, first and last (in this order).

    {{#each items as |item key index first last|}}
    {{/each}}

Also,  when block params are used, no `data` attribute as `@key` or `@index` is generated, so not to conflict with the current context, as no context switching occurs.

    {{#each students}}
        {{#each exams as |exam, examKey|}}
            
            Student name: {{name}} 
            Student key: {{@key}}
            Exam key: {{examKey}}
        {{/each}}
    {{/each}}

I believe creating a new `data` frame while not switching context would break the engine anyway...

**Comment by kpdecker** [comment]
Ember and handlebars have distinct behaviors on many levels. I can't accept
a breaking change like the context one you are proposing. The users will
lynch me if I make breaking changes that are so integral for no reason
other than parity with a project that we never had parity with. I'm glad to
accept smaller changes in distinct PRs but this is too big and too breaking
for me to merge as is.

(Sorry for the quick response, I'm traveling and wanted to provide some
feedback so this doesn't sit without comment for weeks)

**Comment by ericbn** [comment]
Thank you for the quick feedback, @kpdecker.

**Pull request closed without merging**